armv7a9-zynq7000/serial.c

```diff
@@ -0,0 +1,343 @@
+	}
+}
+
+
+static inline void serial_txData(serial_t *serial)
+{
+	/* Keep filling until fifo is not full */
+	while (!(*(serial->base + sr) & (0x1 << 4))) {
+		serial->txHead = (serial->txHead + 1) % BUFFER_SIZE;
+		if (serial->txHead != serial->txTail) {
+			*(serial->base + fifo) = serial->txBuff[serial->txHead];
+			serial->tFull = 0;
+		}
+		else {
+			/* Transfer is finished, disable irq */
+			*(serial->base + idr) |= 1 << 3;
+			break;
+		}
+	}
+}
+
+
+static int serial_irqHandler(u16 n, void *data)
+{
+	u32 st;
+	serial_t *serial = (serial_t *)data;
+
+	if (serial == NULL)
+		return 0;
+
+	st = *(serial->base + isr) & *(serial->base + imr);
+
+	/* RX FIFO trigger irq */
+	if (st & 0x1)
+		serial_rxData(serial);
+
+	/* TX FIFO empty irq   */
+	if (st & (1 << 3))
+		serial_txData(serial);
+
+	/* Clear irq status    */
+	*(serial->base + isr) = st;
+
+	return 0;
+}
+
+
+int serial_read(unsigned int pn, u8 *buff, u16 len, u16 timeout)
+{
+	u16 l, cnt = 0;
+	serial_t *serial;
+	volatile int temp = 0;
+
+	if (pn >= UARTS_MAX_CNT)
+		return ERR_ARG;
+
+	serial = &serial_common.serials[pn];
+
+
+	/* TODO: after adding timer this line should be removed
+	 *       it is necessary to wait some time in order to load kernel via serial */
+	while (serial->rxHead == serial->rxTail && temp++ < 90000)
+		__asm__ volatile ("nop");
+
+	low_cli();
+
+	if (serial->rxHead > serial->rxTail)
+		l = min(serial->rxHead - serial->rxTail, len);
+	else
+		l = min(BUFFER_SIZE - serial->rxTail, len);
+
+	low_memcpy(buff, &serial->rxBuff[serial->rxTail], l);
+	cnt = l;
+	if ((len > l) && (serial->rxHead < serial->rxTail)) {
+		low_memcpy(buff + l, &serial->rxBuff[0], min(len - l, serial->rxHead));
+		cnt += min(len - l, serial->rxHead);
+	}
+	serial->rxTail = ((serial->rxTail + cnt) % BUFFER_SIZE);
+
+	low_sti();
+
+	return cnt;
+}
+
+
+int serial_write(unsigned int pn, const u8 *buff, u16 len)
+{
+	int l, cnt = 0;
+	serial_t *serial;
+
+	if (pn >= UARTS_MAX_CNT)
+		return ERR_ARG;
+
+	serial = &serial_common.serials[pn];
+
+	while (serial->txHead == serial->txTail && serial->tFull)
+		;
+
+	low_cli();
+	if (serial->txHead > serial->txTail)
+		l = min(serial->txHead - serial->txTail, len);
+	else
+		l = min(BUFFER_SIZE - serial->txTail, len);
+
+	low_memcpy(&serial->txBuff[serial->txTail], buff, l);
+	cnt = l;
+	if ((len > l) && (serial->txTail >= serial->txHead)) {
+		low_memcpy(serial->txBuff, buff + l, min(len - l, serial->txHead));
+		cnt += min(len - l, serial->txHead);
+	}
+
+	/* Initialize sending */
+	if (serial->txTail == serial->txHead)
+		*(serial->base + fifo) = serial->txBuff[serial->txHead];
+
+	serial->txTail = ((serial->txTail + cnt) % BUFFER_SIZE);
+
+	if (serial->txTail == serial->txHead)
+		serial->tFull = 1;
+
+	/* Enable TX FIFO empty irq */
+	 *(serial->base + ier) |= 1 << 3;
+	low_sti();
+
+	return cnt;
+}
+
+
+int serial_safewrite(unsigned int pn, const u8 *buff, u16 len)
+{
+	int l;
+
+	for (l = 0; len;) {
```

Comment:
Why not `for (l = 0; len != 0; ++buff, --len) { `?

Reply:
I have based on code:
https://github.com/phoenix-rtos/plo/blob/610d3a55400a1b590bd62e738a70e62d3bff1d76/ia32/msg.c#L42-L53

If it is not readable, I can change it.

Reply:
It is readable, just 2 lines more ;) I leave the decision up to you

Reply:
So I leave it as it is ;)